src/envoy/http/authn/authn_utils.cc

```diff
@@ -36,32 +36,41 @@ bool JsonIterCallback::JsonIteratorCallback(const std::string& key,
   ENVOY_LOG(debug, "{}: key is {}", __FUNCTION__, key);
   ::google::protobuf::Map< ::std::string, ::std::string>* claims =
       payload->mutable_claims();
+
+  // In current implementation, only string objects are extracted into
+  // claims. If call obj.asJsonString(), will get "panic: not reached" from
+  // json_loader.cc.
   try {
+    // Try as string, will throw execption if object type is not string.
+    std::string obj_str = obj.asString();
+    (*claims)[key] = obj_str;
+    // "aud" can be either string array or string.
+    // If it is string, save it to the claims and the payload
+    if (key == "aud") {
+      payload->add_audiences(obj_str);
+    }
+  } catch (Json::Exception& e) {
     if (key == "aud") {
       // "aud" can be either string array or string.
+      // If it is string array, only save it to the payload.
       // Try as string array, read it as empty array if doesn't exist.
-      std::vector<std::string> aud_vector;
       try {
-        aud_vector = obj.getStringArray(key, true);
-      } catch (Json::Exception& e) {
-        // Try as string
-        try {
-          auto audience = obj.asString();
-          aud_vector.push_back(audience);
-        } catch (Json::Exception& e) {
-          ENVOY_LOG(error, "aud field type is not string or string array");
+        std::string json_str = "{\"" + key + "\":" + obj.asJsonString() + "}";
+        Envoy::Json::ObjectSharedPtr json_ptr =
+            Envoy::Json::Factory::loadFromString(json_str);
+        std::vector<std::string> aud_vector =
+            json_ptr->getStringArray(key, true);
+        for (size_t i = 0; i < aud_vector.size(); i++) {
+          ENVOY_LOG(debug, "aud_vector {} is: {}", i, aud_vector[i]);
```

Comment:
Even with debug log, please keep only essential (to debug program). Seeing the value of aud claims here doesn't provide any useful information to see why jwt fail or not fail, and bloat the log a lot.

Reply:
Done.